arctic_training/data/dpo_factory.py

```diff
@@ -0,0 +1,398 @@
+        IGNORE_INDEX
+    ] * len(tokens["prompt_input_ids"])
+    return sequence_tokens
+
+
+class DataCollatorForPref:
+    def __init__(self, tokenizer):
+        self.tokenizer = tokenizer
+
+    def __call__(self, instances: List[Dict]) -> Dict[str, torch.Tensor]:
+        # instances is a list of dictionaries, each dictionary contains:
+        # ['chosen', 'rejected', 'prompt',
+        # 'chosen_input_ids', 'chosen_attention_mask', 'chosen_labels',
+        # 'rejected_input_ids', 'rejected_attention_mask', 'rejected_labels',
+        # 'prompt_input_ids', 'prompt_attention_mask']
+
+        prompt_text = [example["prompt_text"] for example in instances]
+        chosen_text = [example["chosen_text"] for example in instances]
+        rejected_text = [example["rejected_text"] for example in instances]
+
+        input_ids = [
+            torch.tensor(example["chosen_input_ids"]) for example in instances
+        ] + [torch.tensor(example["rejected_input_ids"]) for example in instances]
+        labels = [torch.tensor(example["chosen_labels"]) for example in instances] + [
+            torch.tensor(example["rejected_labels"]) for example in instances
+        ]
+        attention_mask = [
+            torch.tensor(example["chosen_attention_mask"]) for example in instances
+        ] + [torch.tensor(example["rejected_attention_mask"]) for example in instances]
+
+        input_ids = pad(input_ids, padding_value=self.tokenizer.pad_token_id)
+        labels = pad(labels, padding_value=IGNORE_INDEX)
+        attention_mask = pad(attention_mask, padding_value=0)
+
+        rt = {
+                "input_ids": input_ids,
+                "labels": labels,
+                "attention_mask": attention_mask,
+                "prompt": [example["prompt_input_ids"] for example in instances]
+                "prompt_text": prompt_text,
+                "chosen_text": chosen_text,
+                "rejected_text": rejected_text,
+            }
+        return rt
+
+
+@register
+class DPODataFactory(DataFactory):
+    name = "dpo"
+    config: DPODataConfig
+
+    def convert_text(self, tokenizer, conversations: List[Dict[str, str]]):
+        chosen_text = tokenizer.apply_chat_template(
+            conversation=conversations, tokenize=False
+        )
+        return chosen_text
+
+    def process(self, dataset: DatasetType) -> DatasetType:
+        if "prompt" not in dataset.column_names:
+            raise ValueError(
+                "Dataset must have 'prompt' column to tokenize for SFTDataFactory."
+            )
+        if "chosen" not in dataset.column_names:
+            raise ValueError(
+                "Dataset must have 'chosen' column to tokenize for SFTDataFactory."
+            )
+        if "rejected" not in dataset.column_names:
+            raise ValueError(
+                "Dataset must have 'rejected' column to tokenize for SFTDataFactory."
+            )
```

Comment:
```suggestion
        missing_columns = [c for c in ("prompt", "chosen", "rejected") if c not in dataset.column_names]
        if len(missing_columns) > 0:
            raise ValueError(
                f"Dataset must have 'prompt', 'chosen', and 'rejected' columns to tokenizer for DPODataFactory. Missing the following columns: {missing_columns}"
            )
```

This way we can have an error message that indicate if more than one column is missing from the incoming dataset.